### PR TITLE
kvcoord: Fix metrics tracking in mux rangefeed

### LINKED ISSUE #106152: kvcoord/(mux)rangefeed: when changefeed job is cancelled distsender.rangefeed.total_ranges doesn't go down
DistSender maintains a gauge metric showing number of active rangefeeds on the client side. This metric is increased when each feed covering single range is created and decreased when it is cancelled or restarted.

This behaviour seem broken as the gauge doesn't go down if rangefeed is stopped by client by cancelling context. That kind of cancellation happens when changefeed jobs are paused or cancelled.

Note that there may be other effects of rangefeed not being released correctly as this is just manifestation of rangefeed lifecycle management being broken.

**Environment:**
 - CockroachDB version 23.1.4 and likely earlier ones

Jira issue: CRDB-29429

Epic CRDB-26372

### LINKED ISSUE #106252: kvcoord/(mux)rangefeed: when changefeed job is cancelled distsender.rangefeed.catchup_ranges doesn't go down
If changefeed jobs are cancelled or paused while catch up scan is running, metric for those ranges won't be decreased.

Problem is similar to #106152

**Environment:**
 - CockroachDB version 23.1.4 and likely earlier ones

Jira issue: CRDB-29462

Epic CRDB-26372

### PR DESCRIPTION
**Pull request by miretskiy**
Fix an observability bug in mux rangefeed which would incorrectly count various rangefeed related metrics (total ranges, catchup ranges, etc).

Fixes #106152
Fixes #106252

Release note: None

### CODE REVIEW

**Comment by miretskiy** [comment]
@aliher1911 @erikgrinaker 
There is a bit more  change than I hoped, but majority of code changes is to
enable a bit of dependency injection to facilitate the newly added test.

**Comment by aliher1911** [comment]
Looks reasonable. It is unfortunate that we need to run test cluster and all the fragile infra just to test cleanup of client side part. Will dig into details on monday.

File: pkg/kv/kvclient/kvcoord/dist_sender_rangefeed.go [hunk 1]
```diff
@@ -586,18 +593,21 @@ func (a catchupAlloc) Release() {
 }
 
 func acquireCatchupScanQuota(
-	ctx context.Context, ds *DistSender, catchupSem *limit.ConcurrentRequestLimiter,
+	ctx context.Context,
+	sv *settings.Values,
+	catchupSem *limit.ConcurrentRequestLimiter,
+	m *DistSenderRangeFeedMetrics,
```
Comment (erikgrinaker) [inline pkg/kv/kvclient/kvcoord/dist_sender_rangefeed.go:599]:
```suggestion
	metrics *DistSenderRangeFeedMetrics,
```

File: pkg/kv/kvclient/kvcoord/dist_sender_rangefeed.go [hunk 2]
```diff
@@ -89,12 +89,17 @@ func maxConcurrentCatchupScans(sv *settings.Values) int {
 }
 
 type rangeFeedConfig struct {
+	DistSenderRangeFeedMetrics
```
Comment (erikgrinaker) [inline pkg/kv/kvclient/kvcoord/dist_sender_rangefeed.go:92]: nit: this isn't really config, we should pass it separately.

File: pkg/kv/kvclient/kvcoord/dist_sender_rangefeed.go [hunk 3]
```diff
@@ -842,3 +861,26 @@ func (ds *DistSender) handleStuckEvent(
...
+// a test only option to specify metrics to use while executing this rangefeed.
+func withRangeFeedMetrics(m DistSenderRangeFeedMetrics) RangeFeedOption {
+	return optionFunc(func(c *rangeFeedConfig) {
+		c.DistSenderRangeFeedMetrics = m
+	})
+}
+
+// Expose various testing knobs to kvcoord_test package.
+var (
+	TestingWithOnRangefeedEvent = withOnRangefeedEvent
```
Comment (erikgrinaker) [inline pkg/kv/kvclient/kvcoord/dist_sender_rangefeed.go:883]: Why not just export the functions directly?

**Review by miretskiy — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107384)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aliher1911 and @erikgrinaker)

___
*[`pkg/kv/kvclient/kvcoord/dist_sender_rangefeed.go` line 599 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107384#gh-1271908640:-NagJ8iC57zeZRyrK51b:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bf6cd30bf514699be31e678f855d44bf2b7ab231/pkg/kv/kvclient/kvcoord/dist_sender_rangefeed.go#L599)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>


```suggestion
	metrics *DistSenderRangeFeedMetrics,
```
</blockquote></details>

Done.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_rangefeed.go` line 795 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107384#gh-1271912131:-NagJ_3B5SxqOHIfE9hY:b-ool79s) ([raw file](https://github.com/cockroachdb/cockroach/blob/bf6cd30bf514699be31e678f855d44bf2b7ab231/pkg/kv/kvclient/kvcoord/dist_sender_rangefeed.go#L795)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

Ignoring the skip value here violates the knob contract.
</blockquote></details>

Sure; not used w/ regular rangefeed, but no harm in honoring it.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_rangefeed.go` line 865 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107384#gh-1271914052:-NagK2I33Ge2T93XzKob:blq00rb) ([raw file](https://github.com/cockroachdb/cockroach/blob/bf6cd30bf514699be31e678f855d44bf2b7ab231/pkg/kv/kvclient/kvcoord/dist_sender_rangefeed.go#L865)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

Let's move these to where the other options are specified. Also, they're not specific to mux.
</blockquote></details>

ack re mux.  I did want to keep them here because these are test only events.  No need to pollute top level stuff.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_rangefeed.go` line 883 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107384#gh-1271914357:-Nag_cSdCeyfmnUvz31q:b-flvja7) ([raw file](https://github.com/cockroachdb/cockroach/blob/bf6cd30bf514699be31e678f855d44bf2b7ab231/pkg/kv/kvclient/kvcoord/dist_sender_rangefeed.go#L883)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

Why not just export the functions directly?
</blockquote></details>

Okay Still, kept TestingMakeRangeFeedMetrics  alias tough -- I want to be a bit explicit about the fact that it's okay to use in tests only.

**Comment by miretskiy** [comment]
@erikgrinaker  -- ready for another look... back from vaca

**Review by erikgrinaker — APPROVED** [review]
> Reviewed 1 of 5 files at r1, 4 of 4 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107384)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aliher1911)

___
*[`pkg/kv/kvclient/kvcoord/dist_sender_rangefeed.go` line 92 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107384#gh-1271909723:-NajwqoC-Th-hn5KYP9E:bo6etdw) ([raw file](https://github.com/cockroachdb/cockroach/blob/bf6cd30bf514699be31e678f855d44bf2b7ab231/pkg/kv/kvclient/kvcoord/dist_sender_rangefeed.go#L92)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

nit: this isn't really config, we should pass it separately.
</blockquote></details>

Maybe I was wrong about this, I see we do the same thing elsewhere. Your call.

Either way, this comment shouldn't be here.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_rangefeed.go` line 865 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107384#gh-1271914052:-Najxevv2NvEEnSkO-ay:b-xpivv3) ([raw file](https://github.com/cockroachdb/cockroach/blob/bf6cd30bf514699be31e678f855d44bf2b7ab231/pkg/kv/kvclient/kvcoord/dist_sender_rangefeed.go#L865)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

ack re mux.  I did want to keep them here because these are test only events.  No need to pollute top level stuff.
</blockquote></details>

Ok. Consider moving them out to `helpers_test.go` instead, which we normally use for stuff like this -- that also omits them from non-test compilation (which likely doesn't matter here).

**Review by miretskiy — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107384)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aliher1911)

___
*[`pkg/kv/kvclient/kvcoord/dist_sender_rangefeed.go` line 92 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107384#gh-1271909723:-NakRdcl7GrvPafU3OYA:b2im11n) ([raw file](https://github.com/cockroachdb/cockroach/blob/bf6cd30bf514699be31e678f855d44bf2b7ab231/pkg/kv/kvclient/kvcoord/dist_sender_rangefeed.go#L92)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

Maybe I was wrong about this, I see we do the same thing elsewhere. Your call.

Either way, this comment shouldn't be here.
</blockquote></details>

Ack.  I already refactored things, so keeping it.  
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_rangefeed.go` line 865 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107384#gh-1271914052:-NakRj1P-dr3N2dpUfve:bqknc6g) ([raw file](https://github.com/cockroachdb/cockroach/blob/bf6cd30bf514699be31e678f855d44bf2b7ab231/pkg/kv/kvclient/kvcoord/dist_sender_rangefeed.go#L865)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

Ok. Consider moving them out to `helpers_test.go` instead, which we normally use for stuff like this -- that also omits them from non-test compilation (which likely doesn't matter here).
</blockquote></details>

I was thinking about that; alas, things are not that simple.
I cannot put this in helpers_test because helpers_test is in the 'kvcoord' package, while dist_sender_rangefeed_test file is in the `kvcoord_test` package -- thus, I'd need to introduce a new test file for kvcoord_test package.  I cannot move this test file to `kvcoord` package because of circular dependencies.
If I introduce `kvcoord_helpers_test` file to hold those test only only,  then the implementation of those 2 options would need to be able to work on `rangeFeedConfig` -- so that will need to be export; knobs would need to be exported too -- along with its members (or alternatively, I have to introduce public module testing knobs or some such). All of this can be done... it's just it's not clear if it makes things better overall (i.e. the need to export things is a bummer, and everything else I could think of just makes the code more convoluted -- thanks to inability to share the code across two test packages).

**Comment by ericharmeling** [comment]
Do we plan to backport this fix to 22.2? (for context: https://github.com/cockroachlabs/support/issues/2508)

**Comment by miretskiy** [comment]
> Do we plan to backport this fix to 22.2? (for context: https://github.com/cockroachlabs/support/issues/2508)

I don't think 22.2 has this issue.

**Comment by ericharmeling** [comment]
> I don't think 22.2 has this issue.

https://github.com/cockroachlabs/support/issues/2508 (`distsender.rangefeed.catchup_ranges` not reseting) is on `22.2.11`. And it looks the same as https://github.com/cockroachdb/cockroach/issues/106252 (which this PR resolves).

But I could totally be missing something here?

**Comment by miretskiy** [comment]
hmm.. maybe we had this bug earlier...

**Pull request merged**